### ISSUE TITLE
[Minecraft 1.18.2] [Fabric 0.14.9] "io.netty.handler.codec.CorruptedFrameException: Error decoding class com.velocitypowered.proxy.protocol.packet.AvailableCommands"

### ISSUE THREAD

**Issue by unilock** [PaperMC/Velocity#789]
I'm attempting to connect to my Minecraft 1.18.2 server with Fabric 0.14.8 installed (using the modpack [All of Fabric 5](https://www.curseforge.com/minecraft/modpacks/all-of-fabric-5) v1.3.0) through Velocity (velocity-3.1.2-SNAPSHOT-161.jar). Note that I've installed [FabricProxy-Lite](https://modrinth.com/mod/fabricproxy-lite) and [CrossStich](https://modrinth.com/mod/crossstitch) on the server, as well as updated versions of [owo-lib](https://github.com/wisp-forest/owo-lib/issues/37) and [ThonkUtil](https://github.com/LimeAppleBoat/ThonkUtil/issues/25) on both the client and server due to previous incompatibility with Velocity.

Upon attempting to connect, I'm kicked from the server, and the following is printed in Velocity's log:
```
[10:31:18 ERROR]: [server connection] <player_name> -> aof5: exception encountered in com.velocitypowered.proxy.connection.backend.BackendPlaySessionHandler@114d58f8
io.netty.handler.codec.CorruptedFrameException: Error decoding class com.velocitypowered.proxy.protocol.packet.AvailableCommands Direction CLIENTBOUND Protocol 1.18.2 State PLAY ID 12
    at com.velocitypowered.proxy.protocol.netty.MinecraftDecoder.handleDecodeFailure(MinecraftDecoder.java:128) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at com.velocitypowered.proxy.protocol.netty.MinecraftDecoder.tryDecode(MinecraftDecoder.java:83) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at com.velocitypowered.proxy.protocol.netty.MinecraftDecoder.channelRead(MinecraftDecoder.java:58) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:314) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:435) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:279) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:487) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:385) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:995) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at java.lang.Thread.run(Thread.java:833) [?:?]
Caused by: java.lang.IllegalArgumentException: Argument type identifier null unknown.
    at com.velocitypowered.proxy.protocol.packet.brigadier.ArgumentPropertyRegistry.deserialize(ArgumentPropertyRegistry.java:83) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at com.velocitypowered.proxy.protocol.packet.AvailableCommands.deserializeNode(AvailableCommands.java:208) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at com.velocitypowered.proxy.protocol.packet.AvailableCommands.decode(AvailableCommands.java:86) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    at com.velocitypowered.proxy.protocol.netty.MinecraftDecoder.tryDecode(MinecraftDecoder.java:81) ~[velocity-3.1.2-SNAPSHOT-161.jar:3.1.2-SNAPSHOT (git-e8bf6ab5-b161)]
    ... 30 more
```

velocity.toml:
```
# Config version. Do not change this
config-version = "2.0"

# What port should the proxy be bound to? By default, we'll bind to all addresses on port 25577.
bind = "0.0.0.0:25565"

# What should be the MOTD? This gets displayed when the player adds your server to
# their server list. Legacy color codes and JSON are accepted.
motd = "velocity test"

# What should we display for the maximum number of players? (Velocity does not support a cap
# on the number of players online.)
show-max-players = 1337

# Should we authenticate players with Mojang? By default, this is on.
online-mode = true

# Should the proxy enforce the new public key security standard? By default, this is on.
force-key-authentication = true

# If client's ISP/AS sent from this proxy is different from the one from Mojang's
# authentication server, the player is kicked. This disallows some VPN and proxy
# connections but is a weak form of protection.
prevent-client-proxy-connections = false

# Should we forward IP addresses and other data to backend servers?
# Available options:
# - "none":        No forwarding will be done. All players will appear to be connecting
#                  from the proxy and will have offline-mode UUIDs.
# - "legacy":      Forward player IPs and UUIDs in a BungeeCord-compatible format. Use this
#                  if you run servers using Minecraft 1.12 or lower.
# - "bungeeguard": Forward player IPs and UUIDs in a format supported by the BungeeGuard
#                  plugin. Use this if you run servers using Minecraft 1.12 or lower, and are
#                  unable to implement network level firewalling (on a shared host).
# - "modern":      Forward player IPs and UUIDs as part of the login process using
#                  Velocity's native forwarding. Only applicable for Minecraft 1.13 or higher.
player-info-forwarding-mode = "modern"

# If you are using modern or BungeeGuard IP forwarding, configure a file that contains a unique secret here.
# The file is expected to be UTF-8 encoded and not empty.
forwarding-secret-file = "forwarding.secret"

# Announce whether or not your server supports Forge. If you run a modded server, we
# suggest turning this on.
# 
# If your network runs one modpack consistently, consider using ping-passthrough = "mods"
# instead for a nicer display in the server list.
announce-forge = false

# If enabled (default is false) and the proxy is in online mode, Velocity will kick
# any existing player who is online if a duplicate connection attempt is made.
kick-existing-players = false

# Should Velocity pass server list ping requests to a backend server?
# Available options:
# - "disabled":    No pass-through will be done. The velocity.toml and server-icon.png
#                  will determine the initial server list ping response.
# - "mods":        Passes only the mod list from your backend server into the response.
#                  The first server in your try list (or forced host) with a mod list will be
#                  used. If no backend servers can be contacted, Velocity won't display any
#                  mod information.
# - "description": Uses the description and mod list from the backend server. The first
#                  server in the try (or forced host) list that responds is used for the
#                  description and mod list.
# - "all":         Uses the backend server's response as the proxy response. The Velocity
#                  configuration is used if no servers could be contacted.
ping-passthrough = "all"

# If not enabled (default is true) player IP addresses will be replaced by <ip address withheld> in logs
enable-player-address-logging = true

[servers]
# Configure your servers here. Each key represents the server's name, and the value
# represents the IP address of the server to connect to.
vanilla = "127.0.0.1:30000"
mini = "127.0.0.1:30020"
aof5 = "127.0.0.1:30010"

# In what order we should try servers when a player logs in or is kicked from a server.
try = [
  "vanilla"
]

[forced-hosts]
# Configure your forced hosts here.
"play.example.com" = [
  "vanilla"
]
"mini.example.com" = [
  "mini"
]
"aof5.example.com" = [
  "aof5"
]

[advanced]
# How large a Minecraft packet has to be before we compress it. Setting this to zero will
# compress all packets, and setting it to -1 will disable compression entirely.
compression-threshold = 256

# How much compression should be done (from 0-9). The default is -1, which uses the
# default level of 6.
compression-level = -1

# How fast (in milliseconds) are clients allowed to connect after the last connection? By
# default, this is three seconds. Disable this by setting this to 0.
login-ratelimit = 3000

# Specify a custom timeout for connection timeouts here. The default is five seconds.
connection-timeout = 5000

# Specify a read timeout for connections here. The default is 30 seconds.
read-timeout = 30000

# Enables compatibility with HAProxy's PROXY protocol. If you don't know what this is for, then
# don't enable it.
haproxy-protocol = false

# Enables TCP fast open support on the proxy. Requires the proxy to run on Linux.
tcp-fast-open = false

# Enables BungeeCord plugin messaging channel support on Velocity.
bungee-plugin-message-channel = true

# Shows ping requests to the proxy from clients.
show-ping-requests = false

# By default, Velocity will attempt to gracefully handle situations where the user unexpectedly
# loses connection to the server without an explicit disconnect message by attempting to fall the
# user back, except in the case of read timeouts. BungeeCord will disconnect the user instead. You
# can disable this setting to use the BungeeCord behavior.
failover-on-unexpected-server-disconnect = true

# Declares the proxy commands to 1.13+ clients.
announce-proxy-commands = true

# Enables the logging of commands
log-command-executions = false

# Enables logging of player connections when connecting to the proxy, switching servers
# and disconnecting from the proxy.
log-player-connections = true

[query]
# Whether to enable responding to GameSpy 4 query responses or not.
enabled = false

# If query is enabled, on what port should the query protocol listen on?
port = 25565

# This is the map name that is reported to the query services.
map = "velocity"

# Whether plugins should be shown in query response by default or not
show-plugins = false
```

I can't access the client and server logs at the moment, but I don't believe they held anything relevant anyway. Just the standard "player disconnected" messages.

Let me know what more information I can provide.

**Comment by unilock** [comment]
As per the request of @Xernium on Discord, here are the server log and the two(?) dumps made by [`silica-dumper-1.0.2`](https://fivepb.me/drop/silica-dumper-1.0.2.jar).

~~Log: https://pastebin.com/q4pddqtT~~  
~~Dump 1: https://pastebin.com/B3g5Mawv~~  
~~Dump 2: https://pastebin.com/Rhsb8rdX~~

(the `/stop` command was issued from the console manually)

**Comment by Xernium** [comment]
@unilock the paste was marked as spam by pastebin. Please upload the files again either as direct attachment or using https://paste.gg/

**Comment by unilock** [comment]
@Xernium Joyous day!! Here are the files uploaded directly:

[latest.log](https://github.com/PaperMC/Velocity/files/9092534/latest.log)
[dump-server-2022-07-12T09_31_50.749718406Z.txt](https://github.com/PaperMC/Velocity/files/9092537/dump-server-2022-07-12T09_31_50.749718406Z.txt)
[dump-server-2022-07-12T09_31_51.036478497Z.txt](https://github.com/PaperMC/Velocity/files/9092539/dump-server-2022-07-12T09_31_51.036478497Z.txt)

**Comment by unilock** [comment]
It was an issue with Arcanus registering an argument type under the reserved `minecraft:` namespace, which breaks proxies (apparently).
https://github.com/CammiePone/Arcanus/issues/60

As the 1.18 version of Arcanus is no longer supported, a workaround was implemented in CrossStitch.
https://github.com/Xernium/CrossStitch/commit/2530083a0abcd43bcf92c9391065c01e61b97928
([crossstitch-0.1.4-2_build3.jar](https://fivepb.me/drop/crossstitch-0.1.4-2_build3.jar))

A file `crossstitch-overrides.json` must be created in the `config` folder of a given server, specifically on 1.18.* and with Arcanus installed, containing the following:
```
["minecraft:spells","brigadier:spells"]
```
(I assume this general format applies to other mods that incorrectly use reserved namespaces.)

**Comment by unilock** [comment]
Good news, this is happening again! (I posted this on Discord as well, but figured it would be better tracked here.)

My setup has not changed since this issue was closed, aside from updating Velocity and, more importantly, the modpack in question. ([changelog](https://github.com/TeamAOF/All-of-Fabric-5/blob/main/changelogs/changelog_mods_1.4.0.md))

`silica-dumper-1.0.2` dumps + Velocity log: https://paste.gg/p/anonymous/1ef22aa1257a4e57a125bb79d638d83c

**Comment by Neubulae** [comment]
I encountered the same issue, the mods I share with you in common is... Ad Astra and Traveler's Backpack at first glance

**Comment by Neubulae** [comment]
was it [tb](https://github.com/Tiviacz1337/Travelers-Backpack/blob/1.18.2-fabric/src/main/java/com/tiviacz/travelersbackpack/commands/AccessBackpackCommand.java)?

**Comment by Neubulae** [comment]
Well, plain inspection doesn't do anything, so I provide my [dump](https://0x0.st/o_oT.24815988) here too

**Comment by Neubulae** [comment]
...huh, for me it's Tinker's Construct being the culprit, I don't see anything in common on the list at first glance tho

**Comment by Neubulae** [comment]
I tried Compact Machines and it happens just like Tinker's Construct

**Comment by unilock** [comment]
Looks like it is Compact Machines. Same issue as with Arcanus.
https://github.com/AlphaMode/CompactMachines-Fabric/blob/1.18.x/src/main/java/dev/compactmods/machines/CompactMachines.java#L76

I can't speak for Tinker's Construct.

@Xernium, how would one add more entries to `crossstitch-overrides.json`?

**Comment by unilock** [comment]
> ...how would one add more entries to `crossstitch-overrides.json`?

Like this:
```
[
    "minecraft:spells",
    "minecraft:room_pos",
    "minecraft:upgrade_type",
    "brigadier:spells",
    "brigadier:room_pos",
    "brigadier:upgrade_type"
]
```

This fixes AOF5 1.4.0 (and 1.4.1, most likely).

Since the current solution of manually finding the mods that cause this issue and overriding their broken namespaces through CrossStitch isn't ideal, I'll leave this open in case a better solution can be devised. Feel free to close this if that's not necessary or what have you 🤷‍♀️

**Comment by Xernium** [comment]
I’d much rather you report this to the mod authors of the violating mod. That way this issue can be addressed. The overrides are only there for mods that have deemed this a wontfix and the user has to step in.
